### PR TITLE
vim-patch:8.1.{0252,0407,0469,0488,0532,1006,1030,1062,1098,1112,1256} cannot navigate through errors relative to the cursor (take 2)

### PR DESCRIPTION
**Pull request by shanesmith**
Brings over the :cabove, :cbelow, :labove and :lbelow commands introduced in Vim 8.1.1256 (vim/vim#4316).

This is a second attempt at the following PR that grew stale and closed: https://github.com/neovim/neovim/pull/11119

### CODE REVIEW

**Comment by justinmk** [comment]
Thanks! A `src/nvim/testdir/...` test is failing. You can [run that test locally](https://github.com/neovim/neovim/blob/master/test/README.md#running-tests) like this:

    TEST_FILE=test_quickfix.res make oldtest

**Comment by shanesmith** [comment]
I've fixed the linting errors, however the test failure will be trickier...

The first hurdle being that I can't reproduce on my machine (macOS Mojave 10.14.6), the `make oldtest` finishes without any errors.

The second being that C++ isn't my main language and I'm not entirely sure how to tackle the error of `AddressSanitizer: heap-use-after-free`

Any pointers on the above? =)

**Comment by erw7** [comment]
>The second being that C++ isn't my main language and I'm not entirely sure how to tackle the error of AddressSanitizer: heap-use-after-free
>
>Any pointers on the above? =)

I can guess from the following log that the problem is `set_errorlist()`.

```
#0 0x4c5ae2 in __interceptor_free /tmp/final/llvm.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:124:3
#1 0x108c944 in xfree /home/travis/build/neovim/neovim/build/../src/nvim/memory.c:119:3
#2 0x13bd7c5 in ll_free_all /home/travis/build/neovim/neovim/build/../src/nvim/quickfix.c:1704:5
#3 0x13bd1cb in qf_free_all /home/travis/build/neovim/neovim/build/../src/nvim/quickfix.c:1716:5
#4 0x13e9780 in qf_free_stack /home/travis/build/neovim/neovim/build/../src/nvim/quickfix.c:6054:3
#5 0x13e91f1 in set_errorlist /home/travis/build/neovim/neovim/build/../src/nvim/quickfix.c:6089:5
#6 0xa18e6d in set_qf_ll_list /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:15783:7
```

Next, I compared your patch with vim master/HEAD `set_errorlist()`. Then there was a `return OK;` statement after the call to `qf_free_stack()`. Finally, I looked for a commit that contained this change using `git log -p src/quickfix.c`. As a result, I found the next commit that seems to be related to `AddressSanitizer: heap-use-after-free`.

https://github.com/vim/vim/commit/9f84ded38b62c82a4ee57b54f403b1b185ed8170.

**Comment by shanesmith** [comment]
@erw7 thanks for the detective work!

I've now pulled in patch 8.1.0488, let's hope that make CI happy...

**Comment by shanesmith** [comment]
@justinmk @blueyed the tests are green, this PR is ready for your to review =)

File: src/nvim/quickfix.c [hunk 1]
```diff
@@ -1154,35 +1241,259 @@ static char_u * qf_cmdtitle(char_u *cmd)
...
       qi->qf_lists[i - 1] = qi->qf_lists[i];
+    }
     qi->qf_curlist = LISTCOUNT - 1;
   } else
     qi->qf_curlist = qi->qf_listcount++;
-  memset(&qi->qf_lists[qi->qf_curlist], 0, (size_t)(sizeof(qf_list_T)));
-  qf_store_title(qi, qi->qf_curlist, qf_title);
-  qi->qf_lists[qi->qf_curlist].qf_id = ++last_qf_id;
+  qfl = qf_get_curlist(qi);
+  memset(qfl, 0, (size_t)(sizeof(qf_list_T)));
```
Comment (mhinz) [inline src/nvim/quickfix.c]: No need for casting here, since the `sizeof` operator already returns `size_t`.

File: src/nvim/quickfix.c [hunk 2]
```diff
@@ -1154,35 +1241,259 @@ static char_u * qf_cmdtitle(char_u *cmd)
...
+  qf_store_title(qfl, qf_title);
+  qfl->qfl_type = qi->qfl_type;
+  qfl->qf_id = ++last_qf_id;
+}
+
+// Parse the match for filename ('%f') pattern in regmatch.
+// Return the matched value in "fields->namebuf".
```
Comment (mhinz) [inline src/nvim/quickfix.c]: The comments above a function should be Doxygen-style comments, that is `///`. See http://doxygen.nl/manual/docblocks.html#cppblock

File: src/nvim/quickfix.c [hunk 3]
```diff
@@ -80,13 +80,22 @@ struct qfline_S {
 #define LISTCOUNT   10
 #define INVALID_QFIDX (-1)
 
+// Quickfix list type.
+typedef enum
+{
+  QFLT_QUICKFIX,  // Quickfix list - global list
```
Comment (mhinz) [inline src/nvim/quickfix.c]: When commenting members of a struct/union/enum inline, Doxygen-style comments should be used, that is `///<`. See http://doxygen.nl/manual/docblocks.html#memberdoc

File: src/nvim/quickfix.c [hunk 4]
```diff
@@ -792,24 +895,33 @@ static bool qf_stack_empty(const qf_info_T *qi)
...
-  if (qi == NULL || qf_idx < 0 || qf_idx >= LISTCOUNT) {
-    return true;
-  }
-  return qi->qf_lists[qf_idx].qf_count <= 0;
+  return qfl == NULL || qfl->qf_count <= 0;
+}
+
+// Returns TRUE if the specified quickfix/location list is not empty and
+// has valid entries.
+static int qf_list_has_valid_entries(qf_list_T *qfl)
```
Comment (mhinz) [inline src/nvim/quickfix.c]: Opposed to Vim, we use C99. That means we can use the `bool` type as well. This should be `static bool`. Then the comment above should be changed from `TRUE` to `true`.

On compiling you will probably get quite a few warnings that other binary variables etc. using this function should be changed from `int` to `bool` as well.

**Comment by mhinz** [comment]
When the nitpicks about Doxygen-style comments (compare with other functions and structs in `quickfix.c`) and changing `int` to `bool` where appropriate are done, I'll merge this.

**Comment by shanesmith** [comment]
Thanks for the review mhinz! I've addressed your comments in a new commit.

**Review by mhinz — APPROVED** [review]

**Comment by mhinz** [comment]
Thanks! ✨

**Comment by shanesmith** [comment]
Much appreciated!

**Comment by shanesmith** [comment]
I don't mean to be a bother, but not knowing how neovim does releases I was expecting this to be in v0.4.3. Would you be able to give me a hint of when I can see this released? =)

**Comment by justinmk** [comment]
0.4.x is for maintenance releases only. We don't backport inessential things to it. If you don't want to use the development branch, 0.5 will be released in a couple months hopefully.

**Comment by shanesmith** [comment]
oh I see, thanks for the clarification!

**Pull request merged**